### PR TITLE
Fix retrieval of batch indices when dataloader num_workers > 0

### LINKED ISSUE #10782: batch_indices passed to PredictionWriter write_on_epoch_end is wrong when strategy=ddp
## 🐛 Bug

Let's say we are using ddp and there is single dataloader, the number of data points in a process is 140, and the batch size is 64.

When the PredictionWriter's write_on_epoch_end is called on that process, the sizes of predictions and batch_indices parameters are as follows:

```
len(predictions) == 1
len(predictions[0]) == 3
len(predictions[0][0]) == 64
len(predictions[0][1]) == 64
len(predictions[0][2]) == 12

len(batch_indices) == 1
len(batch_indices[0]) == 3
len(batch_indices[0][0]) == 12
len(batch_indices[0][1]) == 12
len(batch_indices[0][2]) == 12
```

Also the contents of `batch_indices[0][0]`, `batch_indices[0][1]`, and `batch_indices[0][2]` are the same. `predictions` are correctly different.

The source of `batch_indices` is `trainer.predict_loop.epoch_batch_indices` which is populated here:

https://github.com/PyTorchLightning/pytorch-lightning/blob/3089dc3829c6456d74ab95aef06891927519eab9/pytorch_lightning/loops/dataloader/prediction_loop.py#L91-L95

`dl_batch_indices` is coming from `PredictionEpochLoop._all_batch_indices`

https://github.com/PyTorchLightning/pytorch-lightning/blob/a002f872eaad5a0f8f2069c49295cdda3568a406/pytorch_lightning/loops/epoch/prediction_epoch_loop.py#L102-L109

Which is filled in the `_store_batch_indices` method:

https://github.com/PyTorchLightning/pytorch-lightning/blob/a002f872eaad5a0f8f2069c49295cdda3568a406/pytorch_lightning/loops/epoch/prediction_epoch_loop.py#L163-L171

During debugging I see `batch_sampler.batch_indices` is in fact the last batch indices every time this method is called.

I couldn't dig deeper but I guess the sampler yields all batches before the program reaches to `_store_batch_indices`.

### Environment

- PyTorch Lightning Version (e.g., 1.5.0): 1.5.3
- PyTorch Version (e.g., 1.10): 1.9.0
- Python version (e.g., 3.9):  3.8
- OS (e.g., Linux):  Linux (Ubuntu)
- CUDA/cuDNN version: 11.3
- GPU models and configuration: 3x RTX 3080
- How you installed PyTorch (`conda`, `pip`, source): Using NVIDIA's Docker Image
- If compiling from source, the output of `torch.__config__.show()`:
- Any other relevant information:

### PR DESCRIPTION
**Pull request by awaelchli**
## What does this PR do?

Fixes  #10782

Please see my comment here https://github.com/PyTorchLightning/pytorch-lightning/issues/10782#issuecomment-983224935 for a detailed answer where the bug comes from and why we need to change the way we use the `IndexBatchSamplerWrapper`. In short, the way a PyTorch BatchSampler retrieves the batch indices differs between `num_workers=0` and `num_workers > 0` in the `DataLoader`, and our prediction loop needs to account for that.

### Does your PR introduce any breaking changes? If yes, please list them.

No.

## Before submitting
- [x] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [x] Did you make sure to update the documentation with your changes? (if necessary)
- [x] Did you write any new necessary tests? (not for typos and docs)
- [ ] Did you verify new and existing tests pass locally with your changes?
- [x] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?
I made sure I had fun coding 🙃

Part of #1 (it's a lie, this is just here to avoid noisy GitHub bot)

cc @borda

### CODE REVIEW

File: pytorch_lightning/loops/epoch/prediction_epoch_loop.py [hunk 1]
```diff
@@ -119,12 +118,10 @@ def _predict_step(self, batch: Any, batch_idx: int, dataloader_idx: int) -> None
         """
         # configure step_kwargs
         step_kwargs = self._build_kwargs(batch, batch_idx, dataloader_idx)
-
-        # extract batch_indices and store them
-        self._store_batch_indices(dataloader_idx)
-
         model_ref = self.trainer.lightning_module
 
+        self.current_batch_indices = self._seen_batch_indices[batch_idx] if self._seen_batch_indices else []
```
Comment (tchaton) [inline pytorch_lightning/loops/epoch/prediction_epoch_loop.py:123]: Won't it be better to pop the `current_batch_indices` ?

Comment (awaelchli) [reply pytorch_lightning/loops/epoch/prediction_epoch_loop.py:123]: I think we need to preserve the full list, as this gets returned `on_run_end` and then returned from `trainer.predict()`. If we don't need the full list, we could pop I guess.

**Review by tchaton — APPROVED** [review]
LGTM !

File: pytorch_lightning/overrides/distributed.py [hunk 2]
```diff
@@ -123,12 +124,31 @@ class IndexBatchSamplerWrapper:
     """This class is used to wrap a :class:`torch.utils.data.BatchSampler` and capture its indices."""
 
     def __init__(self, sampler: BatchSampler) -> None:
+        self.seen_batch_indices: List[List[int]] = []
         self._sampler = sampler
-        self.batch_indices: Optional[List[int]] = None
+        self._batch_indices: List[int] = []
+
+    @property
+    def batch_indices(self) -> List[int]:
```
Comment (awaelchli) [inline pytorch_lightning/overrides/distributed.py:132]: reviewer: the bugfix requires a slight redesign here. This public attribute is kept backward compatible but is not useful anymore, hence the deprecation.

File: tests/callbacks/test_prediction_writer.py [hunk 3]
```diff
@@ -11,54 +11,98 @@
...
-    trainer.predict(model, dataloaders=model.train_dataloader(), return_predictions=False)
-    assert not cb.write_on_batch_end_called
-    assert cb.write_on_epoch_end_called
+    trainer.predict(model, dataloaders=dataloader, return_predictions=False)
+    assert cb.write_on_batch_end.call_count == 0
+    assert cb.write_on_epoch_end.call_count == 1
+
+
+@pytest.mark.parametrize("num_workers", [0, pytest.param(2, marks=RunIf(slow=True))])
+def test_prediction_writer_batch_indices(tmpdir, num_workers):
```
Comment (awaelchli) [inline tests/callbacks/test_prediction_writer.py:85]: reviewer: this is the newly added test that fails on master for num_workers > 0. the test changes above are general improvements and remain true to their original purpose.

**Review by carmocca — APPROVED** [review]

File: pytorch_lightning/overrides/distributed.py (hunk 2 above)
Comment (carmocca) [reply pytorch_lightning/overrides/distributed.py:132]: If this is going into the bugfix release, it should mention the bugfix version it was ddprecated on, not 1.6

Comment (awaelchli) [reply pytorch_lightning/overrides/distributed.py:132]: right, adjusted the range to v1.5.5 - v1.7

File: pytorch_lightning/overrides/distributed.py [hunk 4]
```diff
@@ -123,12 +124,31 @@ class IndexBatchSamplerWrapper:
...
+        rank_zero_deprecation(
+            "The attribute `IndexBatchSamplerWrapper.batch_indices` was deprecated in v1.5.5 and will be removed in"
```
Comment (rohitgr7) [inline pytorch_lightning/overrides/distributed.py:134]:
```suggestion
            "The attribute `IndexBatchSamplerWrapper.batch_indices` was deprecated in v1.5 and will be removed in"
```
I think we just mention v1.x everywhere.

File: pytorch_lightning/overrides/distributed.py [hunk 5]
```diff
@@ -123,12 +124,31 @@ class IndexBatchSamplerWrapper:
...
+            " v1.7. Access the full list `seen_batch_indices` instead."
+        )
+        return self._batch_indices
+
+    @batch_indices.setter
+    def batch_indices(self, indices: List[int]) -> None:
+        rank_zero_deprecation(
+            "The attribute `IndexBatchSamplerWrapper.batch_indices` was deprecated in v1.5.5 and will be removed in"
```
Comment (rohitgr7) [inline pytorch_lightning/overrides/distributed.py:142]:
```suggestion
            "The attribute `IndexBatchSamplerWrapper.batch_indices` was deprecated in v1.5 and will be removed in"
```

File: tests/deprecated_api/test_remove_1-7.py [hunk 6]
```diff
@@ -528,3 +530,12 @@ def is_using_torchelastic():
...
     ):
         MyClusterEnvironment()
+
+
+def test_v1_7_0_index_batch_sampler_wrapper_batch_indices():
+    sampler = IndexBatchSamplerWrapper(Mock())
+    with pytest.deprecated_call(match="was deprecated in v1.5.5 and will be removed in v1.7"):
+        _ = sampler.batch_indices
+
+    with pytest.deprecated_call(match="was deprecated in v1.5.5 and will be removed in v1.7"):
```
Comment (rohitgr7) [inline tests/deprecated_api/test_remove_1-7.py:540]:
```suggestion
    with pytest.deprecated_call(match="was deprecated in v1.5 and will be removed in v1.7"):
        _ = sampler.batch_indices

    with pytest.deprecated_call(match="was deprecated in v1.5 and will be removed in v1.7"):
```

**Review by rohitgr7 — APPROVED** [review]

File: pytorch_lightning/overrides/distributed.py (hunk 4 above)
Comment (awaelchli) [reply pytorch_lightning/overrides/distributed.py]: alrighty almighty!
